mplc/multi_partner_learning.py

```diff
@@ -331,6 +328,81 @@ def fit_minibatch(self):
         logger.debug("End of fedavg collaborative round.")
 
 
+class EnsembleResults(MultiPartnerLearning):
+    def __init__(self, scenario, **kwargs):
+        # First, if only one partner, fall back to dedicated single partner function
+        super(EnsembleResults, self).__init__(scenario, **kwargs)
+        if self.partners_count == 1:
+            raise ValueError('Only one partner is provided. Please use the dedicated SinglePartnerLearning class')
+
+    def evaluate_model(self, x, y_true):
+
+        predictions_list = []
+        for partner in self.partners_list:
+
+            model = partner.build_model()
+            predictions = model.predict(x)
+            predictions_list.append(predictions)
+
+        y_pred = np.mean(predictions_list, axis=0)
+
+        loss = log_loss(y_true, y_pred)
+
+        y_true = np.argmax(y_true, axis=1)
+        y_pred = np.argmax(y_pred, axis=1)
+        metric = accuracy_score(y_true, y_pred)
+
+        hist = [loss, metric]
+
+        return hist
+
+    def fit_epoch(self):
+        # Clear Keras' old models
+        clear_session()
+
+        # Split the train dataset in mini-batches
+        self.split_in_minibatches()
+
+        # Iterate over mini-batches and train
+        for i in range(self.minibatch_count):
+            self.minibatch_index = i
+            self.fit_minibatch()
+
+        self.minibatch_index = 0
+
+    def fit_minibatch(self):
+        """Proceed to a collaborative round with an results ensembling approach"""
+
+        logger.debug("Start new ensembling collaborative round ...")
+        logger.info(f"(ensembling) Minibatch n°{self.minibatch_index} of epoch n°{self.epoch_index}")
+
+        for partner in self.partners_list:
+            partner.model_weights = self.model_weights
+
+        # Evaluate and store accuracy of mini-batch start model
+        self.eval_and_log_model_val_perf()
+
+        # Iterate over partners for training each individual model
+        for partner_index, partner in enumerate(self.partners_list):
+            # Reference the partner's model
+            partner_model = partner.build_model()
+
+            # Train on partner local data set
+            history = partner_model.fit(partner.minibatched_x_train[self.minibatch_index],
+                                        partner.minibatched_y_train[self.minibatch_index],
+                                        batch_size=partner.batch_size,
+                                        verbose=0,
+                                        validation_data=self.val_data)
+
```

Comment:
Since the training of the partners can be done completely separately, I believe it would led to a major speed up to do the whole training at once, in one keras' .fit(), in the way of a single partner training